### PR TITLE
Compile internal "sscanf" module only if needed (#1154)

### PR DESCRIPTION
**Pull request by yonghong-song**
Signed-off-by: Yonghong Song <yhs@fb.com>

### CODE REVIEW

File: src/cc/bpf_module.cc [hunk 1]
```diff
@@ -396,16 +389,27 @@ int BPFModule::annotate() {
...
     }
   }
 
   rw_engine_ = finalize_rw(move(m));
+
+  return 0;
+}
+
+int BPFModule::finalize_annotate() {
+  fn_ptrs_lock_.lock();
```
Comment (mauriciovasquezbernal) [inline src/cc/bpf_module.cc]: What about using a lock guard here to avoid any future problem?

File: src/cc/bpf_module.cc [hunk 2]
```diff
@@ -396,16 +389,27 @@ int BPFModule::annotate() {
...
+  if (has_printf_scanf_) {
+    fn_ptrs_lock_.unlock();
+    return 0;
+  }
+
   if (rw_engine_)
```
Comment (mauriciovasquezbernal) [inline src/cc/bpf_module.cc]: What happens if `rw_engine_` is null? Shouldn't it return at this point?

File: src/cc/bpf_module.cc (hunk 1 above)
Comment (yonghong-song) [inline src/cc/bpf_module.cc]: good suggestion. I tried scoped_lock which is not available for c++11, but I forgot to try lock_guard.

File: src/cc/bpf_module.cc (hunk 2 above)
Comment (yonghong-song) [inline src/cc/bpf_module.cc]: right.

**Comment by yonghong-song** [comment]
@drzaeus77 I hit "Some checks haven't completed yet" for over a day now. Not enough info to see what is the problem? Could you take a look?

**Comment by drzaeus77** [comment]
[buildbot, add to whitelist]

**Comment by yonghong-song** [comment]
test_libbcc uses table scanf/snprintf directly through Table interface. Will add an interface in ebpf::BPF to permit late binding as well.

**Comment by yonghong-song** [comment]
@drzaeus77  Could you take a look?

**Comment by yonghong-song** [comment]
@drzaeus77 just a reminder. Did you get a chance to look at the patch?

**Comment by drzaeus77** [comment]
Yes, I am looking at an alternative.

**Comment by yonghong-song** [comment]
Thanks. I guess you are thinking to reduce scanf/snprintf overhead!

**Comment by drzaeus77** [comment]
Ok, please take a look at #1167 as an alternative. I had several reasons to take this approach: 1. I didn't like the extra std::maps that were used in the code (prior to this commit) 2. llvm::Module already keeps sufficient data structures around, so makes sense to use them (and it has built-in lock guards).

**Pull request closed without merging**